akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1775,10 +1775,17 @@ private[stream] object Collect {
         }
       }
 
-      def pullCondition: Boolean =
-        !strategy.isBackpressure || buffer.used < size
+      def pullCondition: Boolean = strategy match {
+        case EmitEarly =>
+          // when buffer is full we can only emit early if out is available
+          buffer.used < size || isAvailable(out)
```

Comment:
The previous pull-when-full would lead us up to `onPushWhenBufferFull` with an active timer, inside of `onTimer` there is a check to only emit if out is available, which would then lead to the new element being pushed onto the buffer even though it is full, causing all kinds of exciting buffer behavior (but not immediate failure).

Reply:
Thanks for the explanation!